Allow bidirectional communication for TouchOsc

geoffroymontel opened the pull request:
# Purpose
Fixes #313 

# Test case

* start TouchOsc on your iPhone, iPad or Android device and load the Simple layout
* note your device local IP address and incoming port in TouchOSC settings
* on Supercollider

```supercollider
t = MKtl(\tosca, "touchosc-simple1");
t.device.updateSrcAddr("192.168.1.12", 63763); // IP and port of TouchOsc on your device
t.elAt.action = { |el|
    "tOsc %: %\n".postf(el.name, el.value.round(0.001))
};
```

* Move the first slider on the left on the device, you should get values on supercollider console such as 

```
tOsc sl_1: 0.316
```

* On Supercollider, set the value of the first slider with

```supercollider
t.setValueAt(\sl_1, rrand(0.0,1.0));
```

the first slider should move on the device.

LFSaw commented:
Thanks for this!
I'd prefer a separate method `updateDstAddr` so that it is possible to set source and destination to different values.
Alternatively, you can also already now (without touching the code) do:

```
t = MKtl(\tosca, "touchosc-simple1");

// IP and port of TouchOsc on your device
t.device.destination. hostname = "192.168.1.12"; 
t.device.destination. port = 63763; 

t.elAt.action = { |el|
    "tOsc %: %\n".postf(el.name, el.value.round(0.001))
};
```

geoffroymontel commented:
Thanks @LFSaw for the review. I will change it to add a separate `updateDstAddr`.
I could not find a use case where src and dest addresses would be different. Do you have an example ?
Thanks !

LFSaw commented:
If you e.g. use the Pd OSC objects, it is AFAIK not possible to manually set the port from which you're sending.
If this is not desired behaviour (having the possibility to have source and destination different), we should consider refactoring the code accordingly.

> On 24. Oct 2018, at 10:10, Geoffroy Montel <notifications@github.com> wrote:
> 
> Thanks @LFSaw for the review. I will change it to add a separate updateDstAddr.
> I could not find a use case where src and dest addresses would be different. Do you have an example ?
> Thanks !
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

geoffroymontel commented:
Looking at the source code, the `source` parameter is used in the `OSCFunc` to filter incoming OSC messages from the source IP address.

Cf. OSCFun documentation :

```supercollider
OSCFunc.new(func, path, srcID, recvPort, argTemplate, dispatcher)

srcID : An optional instance of NetAddr indicating the **IP address of the sender**. If set this object will only respond to messages from that source.
```

Which means the "port" part of source does not seem to be used at all.

`destination` is used to send back OSC packets to the sender and here, of course, we need both IP and port.

@LFSaw @adcxyz do you think a scenario can exist where source IP is different from destination IP ?
In such a case, I guess merging `source` and `destination` could be a possible refactoring.

geoffroymontel commented:
@LFSaw I followed your suggestion and reverted back `updateSrcAddr` and added `updateDstAddr` and `updateSrcDstAddr`.

I still can't think of a scenario where source IP <> destination IP, if you can't either and you want me to go over refactoring, I'm in !

geoffroymontel commented:
Do you think this PR is ready enough for merge ?

LFSaw commented on Modality/MKtlDescriptions/_allDescs.cache.scd:
```diff
@@ -22,7 +22,6 @@ Dictionary[
 	("evolution-ucontrol-uc33" -> "UC-33 USB MIDI Controller"),
 	("faderfox-uc4" -> "Faderfox UC4"),
 	("faderfox-uc4_pg" -> "Faderfox UC4"),
-	("fio-somo" -> "fio-somo"),
```
why is this file in this PR?

everything else looks good :)

LFSaw commented:
sorry, AFAICS, now you deleted the file `_allDescs.cache.scd` from the repository altogether... what I meant is that you should not add changes to the file to the PR since it is not related to the theme of this PR... (can be a separate PR if needed, though).

LFSaw commented:
Thanks for your patience!

geoffroymontel commented:
Thanks for your detailed review and your patience too !! :)
